server/ServerFSM.cpp

```diff
@@ -521,207 +570,247 @@ sc::result MPLobby::react(const LobbyUpdate& msg) {
+                    // update player connection type for lobby change
+                    player_connection->SetClientType(client_type);
+                } else {
+                    // drop connections for players who were dropped from lobby
+                    m_lobby_data->m_players.erase(player_setup_it);
+                    player_connections_to_drop.push_back(player_connection);
+                }
+            }
 
-        if (psd.m_empire_color != CLR_NONE) {
-            if (!psd_colors.emplace(psd.m_empire_color, entry.first).second) {
-                conflicted_psd_ids.insert(psd_colors.at(psd.m_empire_color));
-                conflicted_psd_ids.insert(entry.first);
+            // drop players connections.  Doing this in separate loop to avoid messing
+            // up iteration above.  these disconnections will lead to Disconnect events
+            // being generated and MPLobby::react(Disconnect) being called.  If this
+            // disconnects the host, then a new host will be selected within that function.
+            for (PlayerConnectionPtr drop_con : player_connections_to_drop) {
+                server.m_networking.Disconnect(drop_con);
             }
-        }
-        /* TODO require unique names, see: MPLobby::react(const JoinGame& msg)
-        if (!psd.m_player_name.empty()) {
-            if (!psd_player_names.emplace(psd.m_player_name, entry.first).second) {
-                conflicted_psd_ids.insert(psd_player_names.at(psd.m_player_name));
-                conflicted_psd_ids.insert(entry.first);
+
+            // remove empty lobby player entries.  these will occur if AIs are dropped
+            // from the lobby.  this will also occur when humans are dropped, but those
+            // cases should have been handled above when checking the lobby data for
+            // each player connection.
+            std::list<std::pair<int, PlayerSetupData>>::iterator player_setup_it = m_lobby_data->m_players.begin();
+            while (player_setup_it != m_lobby_data->m_players.end()) {
+                if (player_setup_it->second.m_client_type == Networking::INVALID_CLIENT_TYPE) {
+                    std::list<std::pair<int, PlayerSetupData>>::iterator erase_it = player_setup_it;
+                    ++player_setup_it;
+                    m_lobby_data->m_players.erase(erase_it);
+                } else {
+                    ++player_setup_it;
+                }
             }
+
         }
-        */
-        if (!psd.m_empire_name.empty()) {
-            if (!psd_empire_names.emplace(psd.m_empire_name, entry.first).second) {
-                conflicted_psd_ids.insert(psd_empire_names.at(psd.m_empire_name));
-                conflicted_psd_ids.insert(entry.first);
+    } else {
+        // can change only himself
+        for (std::pair<int, PlayerSetupData>& i_player : m_lobby_data->m_players) {
+            if (i_player.first == message.SendingPlayer()) {
```

Comment:
Could perhaps reduce the degree of indenting here using `if (!test) continue;` instead of `if (test) {}`